ProjFS.Mac/PrjFSKextTests/MockVnodeAndMount.cpp

```diff
@@ -33,11 +38,100 @@ vnode::vnode(const shared_ptr<mount>& mount) :
 {
 }
 
+vnode::vnode(const std::shared_ptr<mount>& mount, VnodeCreationProperties properties) :
+    mountPoint(mount),
+    name(nullptr),
+    inode(properties.inode == UINT64_MAX ? mount->nextInode++ : properties.inode),
+    type(properties.type),
+    parent(properties.parent)
+{
+}
+
 vnode::~vnode()
 {
     assert(this->ioCount == 0);
 }
 
+static string ParentPathString(const string& path)
+{
+    assert(path.length() > 0);
+    size_t lastSlashPos = path.find_last_of('/');
+    assert(lastSlashPos != string::npos);
+    if (lastSlashPos == 0)
+    {
+        return "/";
+    }
+    else if (lastSlashPos == path.length() - 1) // path ends in "/"
+    {
+        lastSlashPos = path.find_last_of('/', lastSlashPos - 1);
+    }
+    
+    return path.substr(0, lastSlashPos);
+}
+
+// This creates a vnode at the given path below the mount point, and ensures
+// that a hierarchy of directory vnodes exists all the way to the root.
+//
+// For example:
+// mount->CreateVnodeTree("/path/to/file", VREG) will produce:
+// vnode VREG "/path/to/file"
+// vnode VDIR "/path/to"
+// vnode VDIR "/path"
+// vnode VDIR "/"
+// While walking up the path, we stop creating vnodes if one already exists for
+// a particular directory. For example, if the example above is followed by:
+// mount->CreateVnodeTree("/path/for/other/file", VREG) will only generate:
+// vnode VREG "/path/for/other/file"
+// vnode VDIR "/path/for/other"
+// vnode VDIR "/path/for"
+// Because the vnode generated for "/path" in the previous example will be
+// re-used as the parent for the "/path/for" vnode.
+shared_ptr<vnode> mount::CreateVnodeTree(string path, vtype vnodeType)
+{
+    assert(path[0] == '/'); // Only absolute paths allowed
+    shared_ptr<vnode> parentVnode;
+    
+    if (path != "/")
+    {
+        string parentPath = ParentPathString(path);
+        PathToVnodeMap::const_iterator found = s_vnodesByPath.find(parentPath);
+    
+        if (found != s_vnodesByPath.end())
+        {
+            parentVnode = found->second.lock();
```

Comment:
Under what conditions would we find the path, but vnode would null?

Reply:
When the last strong reference (`shared_ptr`) to the vnode goes away, the vnode is destroyed. Any remaining weak references (`weak_ptr`) to that vnode will return null references when attempting to convert them to a strong reference with `lock()`.